orderer/common/throttle/ratelimit.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package throttle
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// Clock observes time
+type Clock interface {
+	Now() time.Time
+	After(d time.Duration) <-chan time.Time
+}
+
+type SharedRateLimiter struct {
+	// Time is how this rate limiter observes time
+	Time Clock
+	// Limit is the total limit of throughput for all clients
+	Limit int
+	// InactivityTimeout is the inactivity time after which a client is pruned from memory
+	InactivityTimeout time.Duration
+	// state
+	once    sync.Once
+	stop    chan struct{}
+	lock    sync.RWMutex
+	clients map[string]*clientState
+}
+
+type clientState struct {
+	lock       sync.Mutex
+	lastSeen   atomic.Value
+	budget     chan struct{}
+	budgetUsed uint32
+}
+
+func (srl *SharedRateLimiter) LimitRate(client string) {
+	srl.once.Do(srl.init)
+
+	cs := srl.getOrCreateState(client)
+
+	srl.lock.RLock()
+	numClients := len(srl.clients)
+	srl.lock.RUnlock()
+
+	budget := srl.Limit / numClients
+
+	budgetWeUsedSoFar := atomic.AddUint32(&cs.budgetUsed, 1)
+
+	if budgetWeUsedSoFar <= uint32(budget) {
+		// We still haven't used up all our budget for the time unit
+		return
+	}
+
+	// Else, budgetWeUsedSoFar > uint32(budget).
+	// We need to:
+	// (1) Lock in order to stall all other goroutines of this client
+	// (2) Wait for a signal that allocates us a new budget portion
+
+	cs.lock.Lock()
+	defer cs.lock.Unlock()
+
+	// Check again, maybe we visited this section from another goroutine?
+	if atomic.LoadUint32(&cs.budgetUsed) <= uint32(budget) {
+		// Some other goroutine already visited this section,
+		// nothing left for us to do.
+		return
+	}
+
+	// Wait for budget to be allocated to us
+	select {
+	case <-cs.budget:
+	case <-srl.stop:
+		return
+	}
+	// Mark budget as allocated and exit
+	cs.lastSeen.Store(time.Now())
+	atomic.StoreUint32(&cs.budgetUsed, 0)
+}
+
+func (srl *SharedRateLimiter) getOrCreateState(client string) *clientState {
+	srl.lock.RLock()
+	cs, exists := srl.clients[client]
+	if exists {
+		srl.lock.RUnlock()
+		return cs
+	}
+	srl.lock.RUnlock()
+
+	srl.lock.Lock()
+	defer srl.lock.Unlock()
+
+	cs, exists = srl.clients[client]
+	if exists {
+		return cs
+	}
+
+	cs = &clientState{
+		budget: make(chan struct{}, 1),
+	}
+
+	cs.lastSeen.Store(time.Now())
+
+	srl.clients[client] = cs
+
+	return cs
+}
+
+func (srl *SharedRateLimiter) Stop() error {
+	select {
+	case <-srl.stop:
+		return fmt.Errorf("stop must be called only once")
+	default:
+		close(srl.stop)
+		return nil
+	}
+}
```

Comment:
Where is the call to this method? Can it be called from 2 different goroutines? If yes, there may be a 2nd channel closure.

Reply:
It's only called in the test. I added a comment about it not being thread safe.